accepted/2020/platform-checks/platform-checks.md

```diff
@@ -0,0 +1,511 @@
+determined by (1).
+
+If (2) is platform-specific it is combined with (1):
+
+* If the platforms between (1) and (2) are disjoint (such as iOS and Windows),
+  the information from (1) is discarded.
+* Otherwise the platform context's minimum version is the maximum of the
+  versions determined by (1) and (2).
+
+### Diagnostics
+
+The analyzer can produce three diagnostics:
+
+1. '{API}' requires {OS} {OS version} or later.
+2. '{API}' has been deprecated since {OS} {OS version}.
+3. '{API}' has been removed since {OS} {OS version}.
+
+### Automatic suppression via platform guards
+
+Diagnostic (1) can be suppressed via a call to
+`RuntimeInformation.IsOSPlatformOrLater()`. This should work for for simple cases
+where the call is contained inside an `if` check but also when the check causes
+control flow to never reach the call:
+
+```C#
+public void Case1()
+{
+    if (RuntimeInformation.IsOSPlatformOrLater(OSPlatform.iOS, 13))
+        AppleApi();
+}
+
+public void Case2()
+{
+    if (!RuntimeInformation.IsOSPlatformOrLater(OSPlatform.iOS, 13))
+        return;
+
+    AppleApi();
+}
+```
+
+Diagnostics (2) and (3) are analogous except the guard is provided by
+`RuntimeInformation.IsOSPlatformEarlierThan()`.
+
+### Code Fixers
+
+* Suggest wrapping statement in platform guard
+* Suggest annotating the method with an attribute
+
+All diagnostics can be suppressed by surround the call site using a platform
+guard. The fixer should offer surrounding the call with an `if` check using the
+appropriate platform-guard. It's not necessary that the end result compiles. It
+is the user's responsibility to adjust control flow and data flow to ensure
+variables are declared and initialized appropriately. The value that the fixer
+provides here is educational and avoids having to lookup the correct version
+numbers.
+
+For diagnostic (1) we should also offer a fixer that allows it to be marked as
+platform specific by applying `AddedInOSPlatformVersionAttribute`. If the
+attribute is already applied, it should offer to change the platform and
+version.
+
+## Q & A
+
+### How does this relate to platform-compat/API Analyzer?
+
+We have an existing project [dotnet/platform-compat], available as the
+[Microsoft.DotNet.Analyzers.Compatibility] NuGet package, but there are several drawbacks:
+
+* The analyzer isn't on by default
+* It carries a database of platform-specific framework APIs, which doesn't scale
+  to 3rd parties
+* Suppression mechanism is somewhat lacking
+
+This is intended to become the productized version of
+[Microsoft.DotNet.Analyzers.Compatibility].
+
+### What set of version numbers are you going to use?
+
+Specifically, are we going to use the "marketing" version numbers or what the
+system returns as version numbers?
+
+This is related to the problem that some operating systems, such as Windows,
+have quirked the OS API to lie to the application in order to make it compatible
+(because way too often apps get version checks wrong).
```

Comment:
That caused everybody to switch to a gradually switch over to different API that is not shimmed... I wound not go into this discussion here. One can argue that this cure was worse than the disease.

Reply:
Fair enough. Let's see what we decide for `Environment.OSVersion` then.